### ISSUE TITLE
[new rule] Disallow !?? (Non-null + nullish)

### ISSUE THREAD

**Issue by fregante** [typescript-eslint/typescript-eslint#2853]
`!` in TypeScript (non-null assertion operator) cancels the null type and `??` (nullish coalescing operator) actually handles it. The first one is redundant. 

This is a suggestion for a new rule:

## Fail

```js
return document.querySelector('a')! ?? defaultElement
```

## Pass

```js
return document.querySelector('a') ?? defaultElement
```
[](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/918#issuecomment-739463405)

**Comment by bradzacher** [comment]
Seems like a logical cousin to `no-non-null-asserted-optional-chain`

**Comment by ypresto** [comment]
It seems covered by `no-unnecessary-conditional`?

![image](https://user-images.githubusercontent.com/400558/131287168-b84bf946-7e77-42d0-8b4b-b527c9bfe267.png)

**Comment by fregante** [comment]
You're right, `no-unnecessary-conditional` was [disabled in my preset](https://github.com/xojs/eslint-config-xo-typescript/blob/c34a940843e5c9e383124364aee64cc2f49a07b7/index.js#L402-L403) due to https://github.com/typescript-eslint/typescript-eslint/issues/1798

**Comment by ypresto** [comment]
It would be appreciated if you can paste the code (example) caused false-positive in your project. Just at assigning array element to variable?

**Comment by fregante** [comment]
There's no false positive. My example matches your example except it uses the same API I used initially.

**Comment by bradzacher** [comment]
Thanks for pointing that out @ypresto!

For reference this could also be handled via `no-restricted-syntax`:
https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#how-can-i-ban-specific-language-feature

```json
{
  "no-restricted-syntax": [
    "error",
     {
       "selector": "LogicalExpression[operator='??'] > TSNonNullExpression.left",
       "message": "Why the heck are you non-nulling a nullish-coalesce?"
     },
  ],
}
```

(dump the config into [here](https://yeonjuan.github.io/typescript-eslint-demo/#e2NvZGU6J3ghID8/IHk7XG54IS55ID8/IHk7XG54ID8/IHkhOyd9) to see it working!)

Closing as there are multiple ways to achieve this without a new rule.

---

For reference - `no-non-null-asserted-optional-chain` makes sense as a rule because it's not achievable with a simple selector alone due to the complexity of the optional chaining AST.

**Comment by fregante** [comment]
> It would be appreciated if you can paste the code (example) caused false-positive in your project

Oh maybe you were referring to this:

- https://github.com/typescript-eslint/typescript-eslint/issues/1798

I didn't personally encounter the issue, I said "my preset" but I meant "the preset I use"
